File: frontend/lib/html-cleaner.ts
/**
 * cleanPostContent — 清洗文章 HTML，确保格式统一
 * ------------------------------------------------------------------
 * 从任意来源（WP 迁移、富文本编辑器、手动粘贴）的 HTML 中剥离
 * 所有内联样式、冗余容器、非标准类名，只保留语义化标签。
 * 清洗后的 HTML 由 .article-body CSS 统一排版。
 */

import sanitizeHtml from "sanitize-html";

// sanitize-html 白名单：只允许文章内容所需的安全标签与属性。
// script/iframe/object 等危险标签、on* 事件属性、javascript:/data: 协议一律被移除（防存储型 XSS）。
const ARTICLE_WHITELIST: sanitizeHtml.IOptions = {
  allowedTags: [
    "h1", "h2", "h3", "h4", "h5", "h6", "p", "a", "ul", "ol", "li", "blockquote",
    "img", "figure", "figcaption",
    "table", "thead", "tbody", "tfoot", "tr", "th", "td", "caption", "colgroup", "col",
    "strong", "b", "em", "i", "u", "s", "mark", "small", "sup", "sub", "code", "pre",
    "hr", "br", "div", "span", "section", "video", "source",
  ],
  allowedAttributes: {
    a: ["href", "name", "target", "rel"],
    img: ["src", "alt", "title", "width", "height", "loading"],
    video: ["src", "controls", "poster", "width", "height"],
    source: ["src", "type"],
    td: ["colspan", "rowspan"],
    th: ["colspan", "rowspan"],
    // 保留 class（.article-body 依赖 alignleft/wp-block-gallery 等类名排版）
    "*": ["class", "id"],
  },
  // 仅允许安全协议，拦截 javascript:/data:/vbscript: 协议注入
  allowedSchemes: ["http", "https", "mailto", "tel"],
  allowedSchemesAppliedToAttributes: ["href", "src"],
  // 外链 target=_blank 自动补 rel="noopener noreferrer"，防 reverse tabnabbing
  transformTags: {
    a: (tagName, attribs) =>
      attribs.target === "_blank"
        ? { tagName, attribs: { ...attribs, rel: "noopener noreferrer" } }
        : { tagName, attribs },
  },
};

/**
 * 清洗所有来源 HTML，确保格式统一
 * - 剥离所有内联 style 属性（格式由 .article-body CSS 接管）
 * - 移除 Word/WP 注入的容器、类名、宽高约束
 * - 移除空段落与 HTML 注释
 */
export function cleanPostContent(html: string): string {
  let cleaned = html;

  // 1. 移除 Astra/WP 主题注入的文章元信息块
  //    包括发布日期、作者、分类等信息行
  cleaned = cleaned.replace(
    /<(?:div|p|span)[^>]*class="[^"]*(?:post-meta|entry-meta|ast-post-meta|posted-on|byline|entry-date|post-info|entry-info)[^"]*"[^>]*>[\s\S]*?<\/(?:div|p|span)>/gi,
    ""
  );

  // 2. 移除 Astra 容器包装：<div class="ast-container">...</div>
  cleaned = cleaned.replace(
    /<div[^>]*class="[^"]*ast-container[^"]*"[^>]*>/gi,
    ""
  );

  // 3. 剥离所有标签上的内联 style 属性（核心规则 — 必须在 width 剥离之前执行）
  //    必须在 width 剥离之前运行：width 正则可能吞噬 style="..." 的闭合引号，
  //    导致后续 style 剥离正则匹配失败。先清掉整段 style="..."，再清残余 width。
  cleaned = cleaned.replace(/\s+style\s*=\s*"[^"]*"/gi, "");
  cleaned = cleaned.replace(/\s+style\s*=\s*'[^']*'/gi, "");

  // 4. 清除内联 width / max-width 样式（残留的独立 width 声明）
  cleaned = cleaned.replace(
    /\s*(?:max-)?width\s*:\s*[^;"]+[;"]?/gi,
    ""
  );

  // 4. 移除 alignwide / alignfull 等 Astra 布局类
  cleaned = cleaned.replace(
    /\b(?:alignwide|alignfull|has-text-align-center|has-background|has-\w+-background-color)\b/gi,
    ""
  );

  // 5. 将 Astra 的 wp-block-cover 等特殊块简化为普通 div
  cleaned = cleaned.replace(
    /<div[^>]*class="[^"]*wp-block-cover[^"]*"[^>]*>/gi,
    '<div class="wp-block-cover">'
  );

  // 6. 移除空的 class 属性
  cleaned = cleaned.replace(/\s+class="\s*"/g, "");
  cleaned = cleaned.replace(/\s+class='\s*'/g, "");

  // 7. 移除 Astra 注入的 <style> 标签
  cleaned = cleaned.replace(
    /<style[^>]*>[\s\S]*?<\/style>/gi,
    ""
  );

  // 8. 移除残留的空 div（配对的 </div></div> 合并为单个）
  cleaned = cleaned.replace(/<div>\s*<\/div>/g, "");

  // 9. 移除 HTML 注释（WP/Astra 注入的不可见标注，如 <!-- BODY CONTENT STARTS -->、<!--more-->）
  //    这些注释对用户不可见，但常残留在空 <p> 内形成无意义占位段，带来额外间距
  cleaned = cleaned.replace(/<!--[\s\S]*?-->/g, "");

  // 10. 移除清洗后仅剩空白的空段落/块（消除正文开头的无意义占位段，
  //     避免其 margin 在图片与正文间制造额外空白）
  cleaned = cleaned.replace(/<(p|div|span)[^>]*>\s*<\/\1>/gi, "");
  cleaned = cleaned.replace(/<(p|div|span)[^>]*>\s*<\/\1>/gi, "");

  // 最后做白名单安全消毒：移除危险标签 / on* 事件属性 / 不安全协议（防存储型 XSS）
  return sanitizeHtml(cleaned, ARTICLE_WHITELIST);
}

File: backend/common/html_cleaner.py
"""HTML 白名单清洗（Shared Kernel，防存储型 XSS）。

设计约束（§4.1 安全设计）：新闻/产品 ``content_html`` 入库前经 bleach 清洗，
仅保留安全标签与属性，防止恶意脚本持久化后在前台执行。

新增 ``clean_text``（security-audit F-01）：标题/摘要/作者等纯文本字段入库前剥离所有
HTML 标签并压缩空白，避免 ``<script>`` 等经 JSON-LD ``dangerouslySetInnerHTML`` 泄露到前台。
"""
from __future__ import annotations

import bleach
import re

# 纯文本字段：剥离所有标签 + 控制字符，压缩空白。
_TEXT_TAG_RE = re.compile(r"<[^>]*>")
_CONTROL_RE = re.compile(r"[\x00-\x08\x0b\x0c\x0e-\x1f\x7f]")

# 允许的标签（常见富文本排版标签）
# 2026-09-18：补充 del / figure / figcaption / mark / s / small —— 后台新闻正文新增「HTML 源码」模式，
# 运营可直接粘贴带图注、删除线、强调语义的 HTML；这几类标签在前台渲染白名单
# （frontend/lib/html-cleaner.ts 的 ARTICLE_WHITELIST）里本就允许，此前只在入库时被剥掉。
ALLOWED_TAGS = [
    "a", "abbr", "b", "blockquote", "br", "code", "del", "div", "em", "figcaption",
    "figure", "h1", "h2", "h3", "h4", "h5", "h6", "hr", "i", "img", "li", "mark",
    "ol", "p", "pre", "s", "small", "span", "strong", "sub", "sup", "table", "tbody",
    "td", "th", "thead", "tr", "u", "ul",
]

# 允许的属性（含图片/链接安全属性）
# security-audit F-06 收敛：通配符仅保留 class，禁止 style 以防 CSS 注入
# （expression()/url() 等向量）。富文本样式统一走 class（Tailwind）。
ALLOWED_ATTRIBUTES = {
    "a": ["href", "title", "target", "rel"],
    "img": ["src", "alt", "title", "width", "height"],
    "*": ["class"],
}

# 允许使用的 URL 协议（防止 javascript: 等）
# tel 与前台渲染白名单（frontend/lib/html-cleaner.ts 的 allowedSchemes）对齐：源码模式粘贴
# 的电话链接此前会被剥掉 href。注意 frontend 白名单里还有 video/source/section 等仅前端
# 允许的标签（历史 WP 内容直入前台的通道），后端不放开这些，属既有的有意差异。
ALLOWED_PROTOCOLS = ["http", "https", "mailto", "tel"]

# 强制为链接添加 rel="noopener"（防 tabnabbing）
ALLOWED_PROTOCOLS_SET = set(ALLOWED_PROTOCOLS)


def clean_html(raw_html: str) -> str:
    """清洗 HTML，去除危险标签/属性/协议。空值安全。"""
    if not raw_html:
        return ""
    cleaned = bleach.clean(
        raw_html,
        tags=ALLOWED_TAGS,
        attributes=ALLOWED_ATTRIBUTES,
        protocols=ALLOWED_PROTOCOLS,
        strip=True,
    )
    # 为外链补全安全属性
    cleaned = bleach.linkify(
        cleaned, callbacks=[bleach.callbacks.target_blank], skip_tags=None
    )
    return cleaned


def clean_text(raw: str | None) -> str:
    """清洗纯文本字段（标题/摘要/作者）：剥离所有 HTML 标签、控制字符，压缩空白。

    security-audit F-01：标题等会进入 JSON-LD 的 ``dangerouslySetInnerHTML``，
    必须以纯文本处理，绝不允许内嵌 HTML/脚本。空值安全，返回字符串。
    """
    if not raw:
        return ""
    # 1) 剥离 HTML 标签
    stripped = _TEXT_TAG_RE.sub("", raw)
    # 2) 去掉控制字符
    stripped = _CONTROL_RE.sub("", stripped)
    # 3) 压缩空白（含全角/半角空格、换行）
    compressed = re.sub(r"\s+", " ", stripped).strip()
    return compressed
